desc/objectives/utils.py

```diff
@@ -121,62 +144,28 @@ def project(x):
 
     def recover(x_reduced):
         """Recover the full state vector from the reduced optimization vector."""
-        dx = put(jnp.zeros(dim_x), unfixed_idx, Z @ x_reduced)
+        dx = put(jnp.zeros(objective.dim_x), unfixed_idx, Z @ x_reduced)
         return jnp.atleast_1d(jnp.squeeze(xp + dx))
 
     # check that all constraints are actually satisfiable
-    xp_dict = {arg: xp[x_idx[arg]] for arg in x_idx.keys()}
+    xp_ = objective.unpack_state(xp, False)
     for con in constraints:
-        res = con.compute_scaled_error(**xp_dict)
-        x = np.concatenate([xp_dict[arg] for arg in con.args])
-        # stuff like density is O(1e19) so need some adjustable tolerance here.
-        if x.size:
-            atol = max(1e-8, np.finfo(x.dtype).eps * np.linalg.norm(x) / x.size)
-        else:
-            atol = 0
+        xpi = [xp_[i] for i, t in enumerate(objective.things) if t in con.things]
+        y1 = con.compute_unscaled(*xpi)
+        y2 = con.target
+        y1, y2 = np.broadcast_arrays(y1, y2)
         np.testing.assert_allclose(
-            res,
-            0,
-            atol=atol,
+            y1,
+            y2,
+            atol=1e-14,
+            rtol=1e-14,
             err_msg="Incompatible constraints detected, cannot satisfy "
             + f"constraint {con}",
         )
 
     return xp, A_full, b_full, Z, unfixed_idx, project, recover
 
 
-def align_jacobian(Fx, objective_f, objective_g):
```

Comment:
Why is this no longer needed?

Reply:
Every objective now takes the full set of arguments, so its jacobian will automatically have zeros in the right places.